### PR TITLE
Add ClientController and ClientService Tests

### PR DESCRIPTION
**Pull request by PatoRocaspana**
### Background
ClientController and ClientService tests were added

### Changes

- xUnit proyect test was added
- ClientController methods were tested implementing Moq library
- ClientService methods were tested

### CODE REVIEW

File: RentACarTests/ControllersTests/ClientControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        #region GetAllAsync
+        [Fact]
+        public async Task GetAllAsync_ReturnsClientsDto()
+        {
+            //arrange
+            var clientOneName = "Pedro";
+
+            var client1 = new Client() { Id = 1, Name = clientOneName, Dni = "13131311" };
+            var client2 = new Client() { Id = 3, Name = "Juanse", Dni = "13131313" };
+            var client3 = new Client() { Id = 7, Name = "Sofia", Dni = "89444666" };
```
Comment (jrocaspana) [inline RentACarTests/ControllersTests/ClientControllerTests.cs]: what's the difference between client1 and client2? why just not create all clients in the same way and assert they were all returned?

File: RentACarTests/ControllersTests/ClientControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            {
+                client1,
+                client2,
+                client3
+            };
+
+            var mockClientService = new Mock<IClientService>();
+            mockClientService.Setup(serv => serv.GetAllAsync()).ReturnsAsync(clientList);
+
+            var clientController = new ClientController(mockClientService.Object);
```
Comment (jrocaspana) [inline RentACarTests/ControllersTests/ClientControllerTests.cs]: this applies to both test classes, I believe you can reutilize this piece of code (except the mock setup) by creating the instance of the mocks and the class to be tested in the test class constructor, let me know if you need some help or guidance

File: RentACarTests/ControllersTests/ClientControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+            //act
+            var resultQuery = await clientController.GetAsync();
+
+            //assert
+            var objectResult = Assert.IsType<OkObjectResult>(resultQuery);
+            var listResult = Assert.IsType<List<ClientDto>>(objectResult.Value);
+            Assert.Empty(listResult);
+            Assert.NotNull(resultQuery);
```
Comment (jrocaspana) [inline RentACarTests/ControllersTests/ClientControllerTests.cs]: following the logical path.. I would first assert the result is not null before verifying if it's of a certain type or not, personally I tend to order the asserts from the most generic one to the most specific ones

File: RentACarTests/ControllersTests/ClientControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            var clientController = new ClientController(mockClientService.Object);
+
+            //act
+            var resultQuery = await clientController.GetAsync();
+
+            //assert
+            var objectResult = Assert.IsType<OkObjectResult>(resultQuery);
+            var listResult = Assert.IsType<List<ClientDto>>(objectResult.Value);
+            Assert.Equal(clientList.Count, listResult.Count);
+            Assert.Equal(clientOneName, listResult.Find(e => e.Id == 1).Name);
```
Comment (jrocaspana) [inline RentACarTests/ControllersTests/ClientControllerTests.cs]: I see on the service tests you verified the mock method was called and also check the params with which it was called. why not do the same here?

File: RentACarWebAPI/Controllers/CarController.cs [hunk 5]
```diff
@@ -58,15 +58,15 @@ public async Task<IActionResult> GetAsync(int id)
...
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
         public async Task<IActionResult> PostAsync([FromBody] CarDto carDto)
         {
             var carEntity = carDto.ToCarEntity(carDto);
 
             var carCreated = await _carService.CreateAsync(carEntity);
 
             if (carCreated is null)
-                return NotFound();
+                return BadRequest();
```
Comment (jrocaspana) [inline RentACarWebAPI/Controllers/CarController.cs:69]: hmm, tricky question: why is this a bad request? not saying the change is bad, just wondering why you changed it

File: RentACarWebAPI/Controllers/RentalController.cs [hunk 6]
```diff
@@ -58,15 +58,15 @@ public async Task<IActionResult> GetAsync(int id)
...
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
         public async Task<IActionResult> PostAsync([FromBody] RentalDto rentalDto)
         {
             var rentalEntity = rentalDto.ToRentalEntity(rentalDto);
 
             var rentalCreated = await _rentalService.CreateAsync(rentalEntity);
 
             if (rentalCreated is null)
-                return NotFound();
+                return BadRequest();
```
Comment (jrocaspana) [inline RentACarWebAPI/Controllers/RentalController.cs:69]: same here, why a bad request?

**Review by jrocaspana — COMMENTED** [review]
nice work! left some questions and suggestions, let me know if you need any help

File: RentACarTests/ControllersTests/ClientControllerTests.cs (hunk 1 above)
Comment (PatoRocaspana) [reply RentACarTests/ControllersTests/ClientControllerTests.cs]: When i started to code this, i had checked if the name of client1 was the same as client1 of the returned list (i was a bit confused in the beginning), so then i changed it and left the same clients. The reason is just anecdotal and I have no excuses! I will change it, thanks!

File: RentACarTests/ControllersTests/ClientControllerTests.cs (hunk 2 above)
Comment (PatoRocaspana) [reply RentACarTests/ControllersTests/ClientControllerTests.cs]: Oh, that is true. I'll make that change and if I get stuck I'll let you know. thanks!

File: RentACarTests/ControllersTests/ClientControllerTests.cs (hunk 4 above)
Comment (PatoRocaspana) [reply RentACarTests/ControllersTests/ClientControllerTests.cs]: You are right. okey!

File: RentACarTests/ControllersTests/ClientControllerTests.cs (hunk 3 above)
Comment (PatoRocaspana) [reply RentACarTests/ControllersTests/ClientControllerTests.cs]: That sounds better! i will change it! Thanks!

File: RentACarWebAPI/Controllers/CarController.cs (hunk 5 above)
Comment (PatoRocaspana) [reply RentACarWebAPI/Controllers/CarController.cs:69]: I think that if the `Post` fails is more probable that the object doesn't comply some requirement or business rule than something `not found`. Honesty, I'm not sure, but it sounds better to me.

Comment (jrocaspana) [reply RentACarWebAPI/Controllers/CarController.cs:69]: a bad request response implies there's something terrible in the request sent from the client, and here what potentially failed is the insert to the DB and _NOT_ the post by itself. I believe the only change `carCreated` would be null is if an exception occurred, and you won't be getting it there

File: RentACarTests/ControllersTests/ClientControllerTests.cs (hunk 1 above)
Comment (jrocaspana) [reply RentACarTests/ControllersTests/ClientControllerTests.cs]: no problem, just try to look for consistency and not to overengineer everything (pretty difficult, I know, but not impossible, just ask for guidance)

File: RentACarWebAPI/Controllers/CarController.cs (hunk 5 above)
Comment (PatoRocaspana) [reply RentACarWebAPI/Controllers/CarController.cs:69]: Ohhh okey, understood! I'll put it back like it was before. Thanks for the explanation and for your time!

Comment (jrocaspana) [reply RentACarWebAPI/Controllers/CarController.cs:69]: no, what was before was "worse" than this. a bad request here would have sense if your service performs some kind of validation and returns null if something isn't right.. if I were you I would return the created object without any null check

**Pull request merged**